controllers/hvpa_controller.go

```diff
@@ -979,7 +986,9 @@ func getWeightedRequests(vpaStatus *vpa_api.VerticalPodAutoscalerStatus, hvpa *a
 					blockedByWeight = true
 
 				} else if diffMem.Sign() > 0 {
-					if hpaScaleOutLimited == false || scaleUpUpdateMode == autoscalingv1alpha1.UpdateModeOff {
+					// Optimization: If we are fully dependent on VPA, then let's wait for HPA scale out to be limited before scaling up
+					// TODO: Should we have a dedicated BlockingReason for this case?
+					if (hpaScaleOutLimited == false && vpaWeight == autoscalingv1alpha1.MaxWeight) || scaleUpUpdateMode == autoscalingv1alpha1.UpdateModeOff {
```

Comment:
Do we need this case now anymore? When would this happen?

Reply:
As mentioned in the comment, this is an optimisation so that vertical scaling happens only when HPA is limited. We can still continue having this, atleast until we tweak hvpa spec for our apiservers. WDYT?

Reply:
Not sure, I understood fully.

Do you mean until we tweak the HVPA spec template for the shoot apiservers in gardener to change the `weightBasedScalingIntervals` to scale vertically first? 

Or do you mean that we should change the HVPASpec type to avoid this case?

Reply:
This:
> until we tweak the HVPA spec template for the shoot apiservers in gardener to change the weightBasedScalingIntervals to scale vertically first

Until we do this or any other modification which is helpful, IMO, we can continue using this optimisation the same way we have been. I think it can be useful, and we can even think of keeping it as an unofficial contract (atleast in the beginning) using alpha annotations. Hint: https://github.com/gardener/gardener/pull/2073

Reply:
> Until we do this or any other modification which is helpful, IMO, we can continue using this optimisation the same way we have been. I think it can be useful, and we can even think of keeping it as an unofficial contract (atleast in the beginning) using alpha annotations. Hint: gardener/gardener#2073

Things were getting clear until the hint part which confused me further :-)

Reply:
I meant that instead of completely removing this optimisation, we can enable it iff some annotation is set on the hvpa resource. Since its a deviation from HVPA controller's definition/contract, we can enable it only if a user has added an annotation. WDYT

Reply:
> Hint: [gardener/gardener#2073](https://github.com/gardener/gardener/pull/2073)

How is the annotation support relevant here?

Reply:
i mentioned this just as an example how Gardener uses these kind of annotations.. Nothing to do with this particular usage.. :)

Reply:
> I meant that instead of completely removing this optimisation, we can enable it iff some annotation is set on the hvpa resource. Since its a deviation from HVPA controller's definition/contract, we can enable it only if a user has added an annotation. WDYT

I am fine keeping it (possibly even without an annotation). But my question was really the following.

How can `hpaScaleOutLimited == false && vpaWeight == autoscalingv1alpha1.MaxWeight` ever be `true` after this PR? Wouldn't that be an error? Am I missing something?

Reply:
Example:
When replicas has reached `maxReplicas` but is not `limited` because HPA wouldn't want to scale out anyway (deducted from `hpa.status.conditions`), but VPA wants to scale up vertically.
Under this `if` condition we block this scale up but only if VPA weight is `MaxWeight`.
It's basically the same behaviour that we have kept for apiserver until now.